### ISSUE TITLE
Blocker - Exception thrown in handleMessage() causing unacked messages on queue

### ISSUE THREAD

**Issue by jwilcox1028** [budjb/grails-rabbitmq-native#57]
Hi Bud,

We are experiencing a situation were messages get stuck in the queue in an unacked and not-ready state.

This is a biggie for us. As we can't have our messages sitting in an unack state on the queue of course and we are 2 weeks from going live. We have some thoughts on a solution for now. Please add your insights asap as you can. We would hate to have to drop your plugin at this point.

Our thoughts on a solution is something like adding a finally clause in the handleMessage() that will force an nack as appropriate, but we have not tried this approach yet, nor looked closer at your code. So any insights you have here would be so valuable to us as we stumble through this.

fyi, we saw your closed issue #54 which we added and appears to at least clear the unacked messages on tomcat start up.

Thanks, Jim (IQNavigator)

Our handleMessage():

```
def handleMessage(def body, MessageContext context) {
    log.info("Notification Consumer started without error")
    NotificationRequest request = new NotificationRequest()
    try {

        request.notificationType = body.notificationType
        request.orgGuid = body.orgGuid
        request.personGuid = body.personGuid
        request.oneTimeUseFlag = body.oneTimeUseFlag
        request.expirationDate = body.expirationDate
        request.entityName = body.entityName
        request.entityGuid = body.entityGuid

        // TODO Replace the following with the proper service call when available
        // notificationTemplateService.handleRequest(request)
        if (request.notificationType == "RESET_PASSWORD")
            notificationSendService.send(request, "Password Reset", "_password_reset")
        else if (request.notificationType == "VERIFY_EMAIL")    
            notificationSendService.send(request, "User Email Validation", "_user_email_verification")

    } catch (Exception e) {
        log.error("Exception processing queued notification: ${request as JSON}", e)
        throw e
    }
    log.info("Notification Consumer completed without error")
}
```

Sample stack trace:

2015-04-24 12:54:31,369 ERROR grails.app.rabbit-consumers.com.iqn.falcon.business.notification.NotificationConsumer {pool-11-thread-2}
[g.a.r.c.i.f.b.n.NotificationConsumer] : Exception processing queued notification
java.lang.NullPointerException: null
        at java.net.URI$Parser.parse(URI.java:3023) ~[na:1.7.0_55]
        at java.net.URI.<init>(URI.java:595) ~[na:1.7.0_55]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.7.0_55]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57) ~[na:1.7.0_55]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.7.0_55]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526) ~[na:1.7.0_55]
        at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:77) ~[groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:10
2) ~[groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57) [groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:182) [groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:190) [groovy-all-2.3.7.jar:2.3.7]
        at grails.plugin.mail.MailMessageContentRenderer$PageRenderRequestCreator.createInstance(MailMessageContentRenderer.groovy:198) ~[MailMessageContentRenderer$PageRenderRequestCreator.class:na]
        at grails.plugin.mail.MailMessageContentRenderer$PageRenderRequestCreator$createInstance.call(Unknown Source) ~[na:na]
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45) [groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108) [groovy-all-2.3.7.jar:2.3.7]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:124) [groovy-all-2.3.7.jar:2.3.7]

**Comment by budjb** [comment]
What version of the plugin are you using where you're experiencing this?

**Comment by jwilcox1028** [comment]
Sorry neglected that detailwe are using this:       compile ":rabbitmq-native:3.1.0"

Version of RabbitMq Server - 3.4.4 (is there a preferable RabbitMq Server version?)

**Comment by budjb** [comment]
Ok, please correct me if I'm wrong, but the general issue seems to be that
unhandled exceptions in the consumer (of the Exception super-type) is not
handled by the plugin, and messages are left un-acked (not ack'd or
nack'd).  Is this correct?  If so, I'll write up a test case and see if I
can reproduce.

**Comment by jwilcox1028** [comment]
that is correct and looked at the plugin code a bit and that seems to be the case.

**Comment by budjb** [comment]
Is the stacktrace you posted complete?  If not, can you provide the rest of it?

**Comment by jwilcox1028** [comment]
We were seeing this issue on one of our servers and I have been trying to reproduce locally and your basicReject is triggering correctly and I have not been able to reproduce so far. I need to reproduce on the server and then will send you a full stack trace or apologize for jumping too quickly on this one.

**Comment by budjb** [comment]
I have a tiny example consumer that has an unhandled exception.  The code follows:

```
package com.budjb.rabbitmq.test

import com.budjb.rabbitmq.consumer.MessageContext
import com.budjb.rabbitmq.publisher.RabbitMessagePublisher

class UnhandledExceptionConsumer {
    RabbitMessagePublisher rabbitMessagePublisher

    /**
     * Consumer configuration.
     */
    static rabbitConfig = [
        consumers: 1,
        queue: 'test.unhandled',
    ]

    /**
     * Handle an incoming RabbitMQ message.
     *
     * @param context Properties of the incoming message.
     * @return
     */
    def handleMessage(MessageContext context) {
        try {
            throw new Exception("I'm unhandled! Wheee!")
        }
        catch (Exception e) {
            throw e
        }
    }
}
```

With this test case, the exception case is appropriately handled and logged by the plugin, and messages are rejected and not stuck in limbo.  Any other information you can provide to reproduce will be helpful.  Let me know!

**Comment by budjb** [comment]
Besides obvious reasons, I asked what version of the plugin you were using specifically because there was an issue in a previous, but recent, version where certain Throwable types were not handled correctly and it left messages unack'd.  This was fixed, but it would have only applied to RuntimeExceptions and Errors.  That was issue #47, and it was fixed in 3.1.0.

**Comment by budjb** [comment]
Have you seen this reoccur?

**Comment by budjb** [comment]
I'm going to close this issue since there's been no more reports of this problem reoccurring.  Feel free to reopen if you come across it again.

**Comment by mdsettles** [comment]
Hi Bud,

Sorry, Jim is no longer on our team and I am just getting around to replying.  Here is the situation we have found that causes the problem.  We are running a Tomcat 7 server.  When we use the hot deploy function, this problem invariably occurs.  If we shutdown tomcat first and then deploy, everything is fine.  If doing the hot deploy, we can also see “old” connections in the RabbitMQ admin console after the deploy is complete and we start publishing messages again.  If we manually force those connections closed, it cleans things up and the new connections start pulling messages correctly.  

So it appears that doing a hot deploy while consumers are running leaves existing connections opened in RabbitMQ, even though they no longer have a consumer running.  RabbitMQ then tries to use the old consumer connections, which obviously will not work.   

For now, we are fine stopping Tomcat to do a deploy.  However, eventually, we are thinking that we need to gracefully shutdown our consumers before doing a hot deploy, which might be a good idea anyway.  Have you ever experienced something similar?  Any suggestions? 
Thanks,
Mike

**Comment by budjb** [comment]
On my team, we always shut down tomcat completely during deployments. We were always running into issues with hot deploys where things did not get cleared our of memory correctly and applications would end up in a weird state and did not function correctly.  I just always assumed this was something weird with Grails and we worked around it.  Due to that, I've never run into this problem.

I've toyed with the idea of a graceful shutdown for a while, but I don't have any concrete plans to implement that functionality.
